src/proxy/mod.rs

```diff
@@ -8,54 +9,61 @@ use hyper::{Method, Request, Response, StatusCode, Uri};
 use hyper_util::rt::TokioIo;
 use std::net::SocketAddr;
 use std::sync::{Arc, Mutex};
+use tokio::io::AsyncWriteExt;
 use tokio::net::{TcpListener, TcpStream};
+use tracing::{debug, info};
 
-use crate::common::types::Backend;
+pub mod backend;
 
-use tokio::sync::Notify;
+use backend::Backend;
 
-#[derive(Debug, Clone)]
+#[derive(Debug)]
 pub struct Server {
     pub listen_addr: SocketAddr,
-    pub backends: Vec<Backend>,
-    current_index: Arc<Mutex<usize>>,
+    pub backends: Vec<Arc<Backend>>,
+    pub current_index: Mutex<usize>,
 }
 
 impl Server {
     pub fn new(listen_addr: SocketAddr, backends: Vec<Backend>) -> Self {
-        Self {
+        let mut new_server = Server {
             listen_addr,
-            backends,
-            current_index: Arc::new(Mutex::new(0)),
+            backends: Vec::new(),
+            current_index: Mutex::new(0),
+        };
+        let mut new_backends = Vec::new();
+
+        for backend in backends.into_iter() {
+            new_backends.push(Arc::new(backend))
         }
```

Comment:
Is there a reason for not passing in a `Vec<Arc<Backend>>` type to this function so you don't have to do this conversion?